### DISCUSSION TITLE
Port Reticulum/apps umbrella to Rust

### DISCUSSION THREAD

**Discussion by hardBSDk** [markqvist/Reticulum#253] [Ideas]
Rust is a powerful language that brings memory safety with performance, consider to use it instead of C (more error-prone).

**Comment by gdt** [comment]
Rust is also nonportable in practice and has a difficult bootstrap story.   But really we need both a protocol spec and an expectation of stability and then various people can write implementations in various languages.

i think it is great that the reference implementation is in python which is broadly portable.

**Comment by NormanMohrmannV** [comment]
Rust is indeed a powerful language that provides memory safety without sacrificing performance. It has been gaining popularity in recent years and is definitely worth considering as an alternative to C, which is known to be more error-prone when it comes to memory safety. In addition, Rust has other features such as built-in concurrency support, zero-cost abstractions, and a rich type system that can help make your code more maintainable and easier to reason about. However, depending on the specific requirements of your project, there may be cases where C is still a more suitable choice. It's important to evaluate your options carefully and choose the best tool for the job.

**Reply by dantee-e** [reply]
I can smell the GPT in this text

**Comment by ghost** [comment]
I dig making code readable and hackable by the masses, but you're combining the two areas begging to prioritize safety the most: network stack (major target for remote exploitation of systems and, specific to Reticulum, control of networks) and cryptography (important secrets that also require heavy computation which impacts packet throughput).

Rust seems the obvious contender in this space, as it's fast and eliminates entire classes of memory vulnerabilities by design. It also boasts strong concurrent execution and async I/O, other important features of the code driving a network stack.

**Reply by markqvist** [reply]
I think these are very valid points. Rust is a great contender for an implementation targeted to general-purpose computers/devices.

Since the Python reference implementation already works here, though, I am not sure it is the most important thing to work on right now. In my eyes, a C implementation working on microcontrollers and embedded devices would be more useful right now.

But, if anyone wants to start a Rust implementation project now, I am absolutely not going to stop them! That would only be great :)

**Reply by gdt** [reply]
> Since the Python reference implementation already works here, though, I am not sure it is the most important thing to work on right now. In my eyes, a C implementation working on microcontrollers and embedded devices would be more useful right now.

I completely agree.  Having played with meshtastic (which has some great aspects but seems fundamentally unsuited to what I want to do), I find the biggest needs in the Reticulum world to be:
  - code for ESP32/LoRa boards (that are sold for meshtastic) to be full Reticulum nodes
  - A way for people with iPhones to be part of the R world.   In the US, a scheme that excludes iPeople is just not viable outside tiny self-selecting communities of purists -- and I say that as someone who agrees with your philosophical points about iOS.  This could be sideband that can talk over wifi to a hotspot provided by the previous bullet, perhaps doing lxmf while proxying R, and being able to pop in for checking messages.   Perhaps it's a webapp from the ESP32.   Perhaps the ESP32 can be set up to beep/blink on message waiting.

**Reply by markqvist** [reply]
When a more fully-featured and self-contained platform is available for MCUs like ESP32, things become a lot more interesting, also in terms of allowing iOS devices to access the network (at least as a sort of "thin client"). With a full C implementation running on small, embedded devices, and a JS implementation for client-side stuff, you can deploy Reticulum anywhere.

It's a bit ironic, since I really dislike the whole idea of JavaScript and web-apps from a conceptual perspective, but in many ways, that path is what makes most sense in terms of massive utility value.

**Comment by ghost** [comment]
I've heard several times now that C should be the language to target microcontrollers and embedded devices. But -- not to sound like too much of a Rust shill -- are you aware that Rust targets that domain too? See https://docs.rust-embedded.org/book/ 

I would understand if the answer is that the tooling is still too immature or something. But I haven't heard anyone's thoughts yet as to why not Rust for embedded devices.

**Reply by markqvist** [reply]
I do know that, but the answer is indeed that the tooling is too immature still ;) Or that I may just not be up to date. But as far as I know, it is still the case. You can't easily write a portable Rust library that just drops right into more or less any toolchain in use right now, for any MCU in production. You can with a C library.

**Reply by npetrangelo** [reply]
I am not sure what the requirements for dropping into toolchains are, but Rust can compile to a lot of different kinds of library formats:
https://doc.rust-lang.org/reference/linkage.html

**Reply by npetrangelo** [reply]
and it can be imported into C
https://www.greyblake.com/blog/exposing-rust-library-to-c/

**Reply by khimaros** [reply]
there has been a lot of progress towards maturing embedded rust since this conversation started. i made use of it for a personal project recently and it is no more difficult to set up than PIO/esp-idf

**Comment by khimaros** [comment]
it looks like this was just released! https://github.com/BeechatNetworkSystemsLtd/Reticulum-rs

**Reply by khimaros** [reply]
quoting their LinkedIn post https://www.linkedin.com/posts/beechat_reticulum-rustlang-opensource-activity-7341079661549944833-q_w7:

> What’s in the Rust port:
> • Full Reticulum protocol, reimplemented from scratch
• std version with a planned no_std update
• Built-in identity, cryptography & routing
• Modular architecture, clean layering
• Runs on real hardware (Kaonic 1S & others)

> 🚧 More coming soon:
> This is just the foundation. We’re actively working on performance optimisations, embedded examples, advanced RF integrations, and cross-platform testing. Expect frequent updates and more tooling to follow

**Reply by pepa65** [reply]
Without effort, this can build statically compiled (musl) binaries that are very portable, and when compressed in the realm of 1MB size (there are 7 binaries provided as examples: `tcp-client`, `kaonic-client`, `kaonic-mesh`, `kaonic-tcp-mesh`, `link-client`, `tcp-server`, and `testnet-client`).

**Reply by RFnexus** [reply]
I would love to test this with the Python version but it doesn't look like it's compatible with 0.9.6 as it's still using AES-128

**Reply by pepa65** [reply]
They claim(ed?) to be fully compatible, so I've tried to contact them to check this out. They are indeed using AES128... When did Reticulum move away from that?

**Reply by markqvist** [reply]
The first release supporting AES-256 was made two months ago, on May 9th, and I informed them on April 15th, that AES-256 support in Reticulum was imminent. Since before they started working on the implementation, they have been aware that Reticulum would move to AES-256 at some point, though. So I guess they just haven't gotten around to adding it yet. Should be very simple to do though, it's really not much of a code change, especially now that the implementation here in the reference implementation is finalized.

**Reply by pepa65** [reply]
Indeed, the change is trivial. I build the (statically compiled) binaries with AES256, available for download here: https://github.com/pepa65/reticulum-rs/releases/tag/0.2.1
I'd be curious to know if they worked.

**Reply by markqvist** [reply]
Aweomse job man!

**Reply by pepa65** [reply]
Are they usable??

**Reply by markqvist** [reply]
Won't be able to test it myself for some time, I'll be off for an extended break in a day or two, so hopefully someone else can help test it out :)

**Reply by n-beechat** [reply]
@pepa65  thanks for your comments we have now added AES256
https://github.com/BeechatNetworkSystemsLtd/Reticulum-rs/commit/d86eb2ed27e946a8a7df2136464d2303e85f1a86

**Reply by helmick43270** [reply]
Running Sideband 1.7.0 for Android, 1.7.0 for Linux, and tcp-server.rs using TCP interface only. Voice and text messages go back and forth like normal.

**Reply by helmick43270** [reply]
tcp-server builds and runs on Termux, too! 3.6 MB binary. Same functionality as above.

**Reply by kujeger** [reply]
I gave the tcp-server a shot too; works great nomadnet 0.8.0 <-> sideband 1.7.0 <-> meshchat 2.1.0 on three devices all connecting to the same tcp-server. Pretty cool!

**Reply by pepa65** [reply]
It can be a statically compiled binary under 1 MB, which is very cool!

**Reply by helmick43270** [reply]
This afternoon I’d like to try the 999kB binary you built. It looks like it’s slightly different from the beechat build.

**Reply by David-OConnor** [reply]
Would these be adjustable to work on embedded/no-std/no alloc? I suppose I am confused about Reticulum, as I would think this is a no-std-primary tool, as it's a network for the sort of radio that goes on embedded devices. Ty!

**Reply by pepa65** [reply]
> This afternoon I’d like to try the 999kB binary you built. It looks like it’s slightly different from the beechat build.

It should functionally be the same, but it is statically compiled with musl, and compressed.

**Reply by helmick43270** [reply]
Yes, it does well with text and voice. Was curious if tcp-server would send images. I raised the issue a few days ago on the Beechat site. Didn't know if it was something broken, or just not there yet.

**Comment by David-OConnor** [comment]
Hey! Also interested in using Reticulum in Rust, but I think this is burying the lede a bit... I think the more practical topic is *Using Reticulum on arbitrary hardware and software*. The shape of the solution is *What bytes do I read and write at what time*, so maybe we move towards more software-agnostic docs and mindset? I suspect this is what @gdt 's comment at the top of this thread is getting at.

Example: That reticulum-rs lib looks... not very practical as it requires std rust. It also requires tokio, and an allocator. This is confusing to me, since the sorts of radios of interest here are more (unless I am missing something) more relevant to no-std applications.

**Reply by helmick43270** [reply]
Could be a little wobble in the topic, given the two-year lull in this discussion, then the recent interest in someone’s completion of a Rust incarnation of Reticulum. I don’t really know much about this kind of thing, and that’s why I’m here, to learn about it. I think you’re right about the essence, to eventually get Reticulum to run on anything. I’m trying to get it going in legacy Termux on Android 5 on a cheap POC radio. That’s what I do right now to relax.

**Reply by helmick43270** [reply]
tcp-server runs on legacy Termux! I didn't try building it there.

**Reply by kujeger** [reply]
while tokio is a bit much for embedded, no-std isn't necessarily a requirement. esp32's have the esp-idf from espressif that enables building std things:
https://docs.espressif.com/projects/rust/book/overview/using-the-standard-library.html

**Reply by pepa65** [reply]
Tokio makes is async, so that's what that's for, useful for embedded too. And it doesn't make it large (or bloated).

**Reply by kujeger** [reply]
For sure, tokio and async are absolutely useful!
It's just that afaik, tokio isn't much used in embedded devices for async, since it requires std -- for example, the embassy framework is more often used. But I do see that some people have tokio running on esp32, I didn't actually know that!

**Reply by n-beechat** [reply]
Currently we're focusing on std for now and will do no_std as part of a later migration

**Reply by kujeger** [reply]
just for fun I did get an esp32-c3 up and running with tokio, and even starting reticulum-rs, but it immediately runs out of memory and explodes :sweat_smile:

**Reply by pepa65** [reply]
The python version works??

**Reply by helmick43270** [reply]
Referring to the ESP32-C3 above, or the issue on tcp_server.rs? Concerning the issue, yes, images do fine on the python version.

**Reply by pepa65** [reply]
If the Rust version doesn't really run on the ESP32-C3, does the python version run okay and not run out of memory??

**Reply by kujeger** [reply]
the python version will not run on any embedded devices, no -- it also requires much more memory etc. This is one of the reasons a rust (and zig, and c(++)) reimplementation of reticulum is so exciting!

**Reply by helmick43270** [reply]
@pepa65, I've been running your compressed tcp-server build on a remote server, rnsrsusa.vps.webdock.cloud, port 4242. Seems to do okay as long as the payload is no larger than 2048 bytes. That's the way the beechat build was doing on it, too. The Webdock server is the absolute smallest, cheapest selection. It's running alongside a different service, and it barely moves the htop gauge.

**Reply by helmick43270** [reply]
With maximum payload of ~2kB, that’s a voice (codec2) message of somewhere around 5 seconds max. It appears that connections have been established, but same errors reported when I tried images or voice messages of size approaching 2kB.
